tgg/relatedness/n_non_ref_pairs.py

```diff
@@ -0,0 +1,273 @@
+from ukbb_qc.resources.sample_qc import interval_qc_path
+from ukbb_qc.resources.variant_qc import info_ht_path, NA12878, SYNDIP
+
+logging.basicConfig(
+    format="%(asctime)s (%(name)s %(lineno)s): %(message)s",
+    datefmt="%m/%d/%Y %I:%M:%S %p",
+)
+logger = logging.getLogger("n_non_ref_utils")
+logger.setLevel(logging.INFO)
+
+
+TEMP_PATH = "gs://gnomad-tmp/n_non_ref"
+"""
+Path to bucket for temporary data.
+"""
+
+VDS_PATH = "gs://ukbb-pharma-exome-analysis/500k_temp/500k.vds"
+"""
+Path to Variant Dataset (VDS) that contains high quality samples from the final UK Biobank callset.
+
+Generated using the following commands:
+
+# NOTE: If ever need to rerun, should read v4 VDS using `get_gnomad_v4_vds`
+# (in gnomad_qc/v4/resoures/basics.py)
+vds = hl.vds.read_vds("gs://gnomad/raw/exomes/4.0/gnomad_v4.0.vds")
+meta_ht = hl.read_table("gs://broad-ukbb/broad.freeze_7/sample_qc/meta.ht")
+meta_ht = meta_ht.filter(meta_ht.sample_filters.high_quality)
+vds = hl.vds.filter_samples(vds, meta_ht, remove_dead_alleles=True)
+call_stats_ht = hl.read_table(ukb.var_annotations_ht_path('ukb_freq', *TRANCHE_DATA[CURRENT_TRANCHE]))
+freq_index = get_cohort_index(call_stats_ht)
+var = vds.variant_data.annotate_rows(call_stats=call_stats_ht[vds.variant_data.row_key].freq[freq_index])
+vds = hl.vds.VariantDataset(vds.reference_data, var)
+vds.write(ukb_exomes_path, overwrite=args.overwrite)
+
+Full VDS size: 8.69 TiB
+VDS variant data size: 561.97 GiB
+"""
+
+TRANCHE_DATA = ("broad", CURRENT_FREEZE)
+"""
+UKB tranche data (data source and data freeze number)
+"""
+
+
+def get_n_non_ref_sites(
+    vds_path: str = VDS_PATH,
+    temp_path: str = TEMP_PATH,
+    tranche_data: Tuple[str, int] = TRANCHE_DATA,
+    autosomes_only: bool = True,
+    snp_only: bool = False,
+    het_only: bool = False,
+    adj_only: bool = True,
+    interval_qc_regions: bool = True,
+    no_AS_lowqual: bool = True,
+    non_ref_samples: int = 3,
+) -> hl.Table:
+    """
+    Filter VDS to autosomal sites in interval QC pass regions with an adj non ref of n.
+    :param vds_path: Path to VDS. Default is VDS_PATH.
+    :param temp_path: Path to bucket to store Table and other temporary data. Default is TEMP_PATH.
+    :param tranche_data: UKB tranche data (data source and data freeze number). Default is TRANCHE_DATA.
+    :param autosomes_only: Filter VDS to autosomes. Defaults to True.
+    :param snp_only: Filter VDS to bi-allelic SNPs. Defaults to False.
+    :param het_only: Filter to only het samples. Defaults to False.
+    :param adj_only: Filter GT to adj. Defaults to True.
+    :param interval_qc_regions: Filter to interval QC regions. Defaults to True.
+    :param no_AS_lowqual: Remove AS_lowqual sites. Defaults to True.
+    :param non_ref_samples: Number of non ref samples found in each variant to filter to, e.g. for tripletons, n_samples=3. Defaults to 3.
+    :return: Table of high quality sites with private to n samples.
```

Comment:
```suggestion
    :param non_ref_samples: Desired number of non ref samples for each variant, e.g. for tripletons, n_samples=3. Defaults to 3.
    :return: Table of high quality sites filtered to variants with specified number of non-reference samples.
```
(not sure how to make this less wordy)